language/compiler/bytecode_source_map/src/source_map.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use failure::prelude::*;
+use ir_to_bytecode_syntax::ast::{Loc, QualifiedModuleIdent, Var_};
+use libra_types::account_address::AccountAddress;
+use libra_types::identifier::Identifier;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::ops::Bound;
+use vm::file_format::{
+    CodeOffset, FieldDefinitionIndex, FunctionDefinitionIndex, StructDefinitionIndex, TableIndex,
+};
+use vm::internals::ModuleIndex;
+
+//***************************************************************************
+// Source location mapping
+//***************************************************************************
+
+pub type SourceMap = Vec<ModuleSourceMap>;
+pub type SourceName = (Identifier, Loc);
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct StructSourceMap {
+    // The source declaration location of the struct
+    pub decl_location: Loc,
+
+    // NB type parameters need to be added in the order of their declaration
+    pub type_parameters: Vec<SourceName>,
+
+    // NB that fields to a struct source map need to be added in the order of the fields.
+    pub fields: Vec<Loc>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct FunctionSourceMap {
+    // The source location for the definition of this entire function. Note that in certain
+    // instances this will have no valid source location e.g. the "main" function for modules that
+    // are treated as programs are synthesized and therefore have no valid source location.
+    pub decl_location: Loc,
+
+    // NB type parameters need to be added in the order of their declaration
+    pub type_parameters: Vec<SourceName>,
+
+    // The index into the vector is the locals index. The corresponding `(Identifier, Loc)` tuple
+    // is the name and location of the local.
+    pub locals: Vec<SourceName>,
+
+    // The source location map for the function body.
+    pub code_map: BTreeMap<CodeOffset, Loc>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ModuleSourceMap {
+    // The name <address.module_name> for module that this source map is for
+    pub module_name: (AccountAddress, Identifier),
+
+    // A mapping of StructDefinitionIndex to source map for each struct/resource
+    struct_map: HashMap<TableIndex, StructSourceMap>,
```

Comment:
Nit: can we make these into `BTreeMap`s? If someone iterated over the marks, they could get different orderings in different iters. This might or might not matter, but seems good to avoid the surpris if we can.